rust/datafusion/src/physical_plan/distinct_expressions.rs

```diff
@@ -0,0 +1,303 @@
+        Some(ScalarValue::Int16(v)) => Some(DistinctScalarValue::Int16(v)),
+        Some(ScalarValue::Int32(v)) => Some(DistinctScalarValue::Int32(v)),
+        Some(ScalarValue::Int64(v)) => Some(DistinctScalarValue::Int64(v)),
+        Some(ScalarValue::UInt8(v)) => Some(DistinctScalarValue::UInt8(v)),
+        Some(ScalarValue::UInt16(v)) => Some(DistinctScalarValue::UInt16(v)),
+        Some(ScalarValue::UInt32(v)) => Some(DistinctScalarValue::UInt32(v)),
+        Some(ScalarValue::UInt64(v)) => Some(DistinctScalarValue::UInt64(v)),
+        Some(ScalarValue::Null) => None,
+        _ => {
+            return Err(ExecutionError::NotImplemented(
+                "Unsupported scalar value for DISTINCT accumulator".to_string(),
+            ))
+        }
+    };
+
+    match accum_value {
+        Some(v) => {
+            accum.insert(v);
+        }
+        None => {}
+    }
+
+    Ok(())
+}
+
+macro_rules! accum_batch {
+    ($ARRAY_TY:ident, $DISTINCT_SCALAR_TY: path, $ARRAY: expr, $ACCUM: expr) => {{
+        let array = $ARRAY.as_any().downcast_ref::<$ARRAY_TY>().ok_or_else(|| {
+            ExecutionError::ExecutionError("Error downcasting array".to_string())
+        })?;
+
+        for i in 0..array.len() {
+            $ACCUM.insert($DISTINCT_SCALAR_TY(array.value(i)));
+        }
+
+        Ok(())
+    }};
+}
+
+#[derive(Debug)]
+struct DistinctValuesAccumulator {
+    values: FnvHashSet<DistinctScalarValue>,
+}
+
+impl Accumulator for DistinctValuesAccumulator {
+    fn accumulate_scalar(&mut self, value: Option<ScalarValue>) -> Result<()> {
+        accumulate_scalar(&mut self.values, value)
+    }
+
+    fn accumulate_batch(&mut self, _array: &ArrayRef) -> Result<()> {
+        Err(ExecutionError::NotImplemented(
+            "Aggregates with DISTINCT not supported without a GROUP BY".to_string(),
+        ))
+    }
+
+    fn get_value(&self) -> Result<Option<ScalarValue>> {
+        let value_out = self
+            .values
+            .iter()
+            .map(|accumulated_value| match accumulated_value {
+                DistinctScalarValue::Int8(v) => Ok(ScalarValue::Int8(*v)),
+                DistinctScalarValue::Int16(v) => Ok(ScalarValue::Int16(*v)),
+                DistinctScalarValue::Int32(v) => Ok(ScalarValue::Int32(*v)),
+                DistinctScalarValue::Int64(v) => Ok(ScalarValue::Int64(*v)),
+                DistinctScalarValue::UInt8(v) => Ok(ScalarValue::UInt8(*v)),
+                DistinctScalarValue::UInt16(v) => Ok(ScalarValue::UInt16(*v)),
+                DistinctScalarValue::UInt32(v) => Ok(ScalarValue::UInt32(*v)),
+                DistinctScalarValue::UInt64(v) => Ok(ScalarValue::UInt64(*v)),
+            })
+            .collect::<Result<Vec<ScalarValue>>>()?;
+
+        Ok(Some(ScalarValue::Struct(value_out)))
+    }
+}
+
+/// Create a distint count expression.
+pub fn count(mode: AggregateMode, expr: Arc<dyn PhysicalExpr>) -> Arc<dyn AggregateExpr> {
```

Comment:
Note that this will not apply any coercion to the arguments. We are trying to make `aggregates::create_aggregate_expr` be the entry point to create the physical expressions, to guarantee that the signature is correct and coercion rules apply.

However, I see that this depends on the `AggregateMode`, which the current `create_aggregate_expr` does not support. #8172 addresses exactly this. :)

Reply:
Good catch, thanks for pointing that out. I'll have a look at this as soon as I integrate with #8172.